### DISCUSSION TITLE
Question regarding select_only

### DISCUSSION THREAD

**Discussion by elderapo** [SeaQL/sea-orm#731] [Q&A]
Hey, I am new to rust so please excuse my lack of knowledge about specific rust things.

I was wondering if it was possible to somehow do:
```rs
let result_with_username_and_age_only = user::Entity::find_by_id(2)
    .select_only()
    .column(user::Column::Username)
    .column(user::Column::Age)
    .into_model() // <-- return type inferred based on column/columns specified above
    .one(&db)
    .await
    .unwrap()
    .unwrap();
```
instead of
```rs
#[derive(Debug, FromQueryResult)]
struct UsernameResult {
    username: String,
}

let result_with_username_and_age_only = user::Entity::find_by_id(2)
    .select_only()
    .column(user::Column::Username)
    .column(user::Column::Age)
    .into_model::<UsernameResult>()
    .one(&db)
    .await
    .unwrap()
    .unwrap();
```

it would shorten the user-land code for custom selects quite a bit and add additional type-safety (ex. when selecting `username` column it's possible to make a typo in the result struct and write `usernamea`). This runtime casting makes sense for raw queries (it would be pretty hard to figure out result type based just on SQL query) but here I think the result type could be based on columns specified in `.column(...)` calls which already ensure to not select columns that do not exist (at least on entity-level - entity still could be unsynchronized from an actual database schema).

**Comment by tyt2y3** [comment]
It is possible, if we attach a derive macro to `UsernameResult` and have it impl a `CustomSelect` trait, then it can be handled directly by the `Select`.
It is moderately difficult though, because it involves derive macros.

**Reply by billy1624** [reply]
I think what @elderapo suggested is that we don't need to manually define... at all.

```rs
#[derive(Debug, FromQueryResult)]
struct UsernameResult {
    username: String,
}
```

Calling `into_model` would infer the returning struct and all of its field types implicitly

```rs
.into_model() // <-- return type inferred based on column/columns specified above
```

**Reply by penso** [reply]
Yes that'd be great.

**Reply by elderapo** [reply]
> I think what @elderapo suggested is that we don't need to manually define... at all.
> 
> 
> 
> ```rs
> 
> #[derive(Debug, FromQueryResult)]
> 
> struct UsernameResult {
> 
>     username: String,
> 
> }
> 
> ```
> 
> 
> 
> Calling `into_model` would infer the returning struct and all of its field types implicitly
> 
> 
> 
> ```rs
> 
> .into_model() // <-- return type inferred based on column/columns specified above
> 
> ```

Thats exacly what I meant.

**Comment by penso** [comment]
I need exactly this to be able to loop through a dynamic list of columns, when using `async_graphql` and lookahead to know what needs to be fetched. Is there any update on that?

Sadly I'm not sure I'm proficient enough in Rust to write that macro myself.

**Reply by billy1624** [reply]
Hey @penso, what's the plan to hold the dynamic result? Seems to me it's not possible to define all combination of custom struct to hold the result.

You might want to `try_get` each column by your own, given that you know the corresponding Rust type of each column

```rs
let id: i32 = query_res.try_get("", "id")?;
```

https://www.sea-ql.org/SeaORM/docs/basic-crud/raw-sql#get-custom-result-using-query_one-and-query_all-methods

**Reply by penso** [reply]
I already do that using sqlx but I was hoping to use sea-orm and not a mix of both based on what I need to do. The dynamic result would be send over graphql with `async_graphql`. But maybe that wouldn't work since the returned type would change based on what columns were requested. Maybe I'll need a Struct type with all columns listed, as optional.

**Reply by billy1624** [reply]
Is it possible to return a serialized struct to the GraphQL client? Such as a JSON

Then, we can use the `into_json` API provided by SeaORM. To construct the JSON dynamically: https://www.sea-ql.org/SeaORM/docs/basic-crud/json

**Reply by penso** [reply]
I don't think you can because `async_graphql` uses the return type to generate the GraphQL schema. I don't see how that could work.
